File: server/src/models/Document.ts
import mongoose, { Schema, Document } from 'mongoose';

export interface IDocument extends Document {
  name: string;
  type: string; // ex: 'NR-11', 'CNH', 'Certificado', 'Manual'
  category: 'operator' | 'equipment';
  ownerId: mongoose.Types.ObjectId; // ID do operador ou equipamento
  fileUrl: string;
  fileName: string;
  mimeType: string;
  size: number;
  expiresAt?: Date;
  status: 'valid' | 'warning' | 'expired';
  createdAt: Date;
  updatedAt: Date;
}

const DocumentSchema: Schema = new Schema(
  {
    name: { type: String, required: true },
    type: { type: String, required: true },
    category: { type: String, enum: ['operator', 'equipment'], required: true },
    ownerId: { type: Schema.Types.ObjectId, required: true },
    fileUrl: { type: String, required: true },
    fileName: { type: String, required: true },
    mimeType: { type: String, required: true },
    size: { type: Number, required: true },
    expiresAt: { type: Date },
    status: {
      type: String,
      enum: ['valid', 'warning', 'expired'],
      default: 'valid',
    },
  },
  { timestamps: true }
);

// Middleware para atualizar status baseado na data de validade antes de salvar
DocumentSchema.pre<IDocument>('save', function (next) {
  if (this.expiresAt) {
    const now = new Date();
    const thirtyDaysFromNow = new Date();
    thirtyDaysFromNow.setDate(now.getDate() + 30);

    if (this.expiresAt < now) {
      this.status = 'expired';
    } else if (this.expiresAt < thirtyDaysFromNow) {
      this.status = 'warning';
    } else {
      this.status = 'valid';
    }
  }
  next();
});

export default mongoose.model<IDocument>('Document', DocumentSchema);

File: mobile/lib/models/document.dart

class DocumentModel {
  final String id;
  final String name;
  final String type;
  final String category;
  final String ownerId;
  final String fileUrl;
  final String fileName;
  final String mimeType;
  final int size;
  final DateTime? expiresAt;
  final String status;
  final DateTime createdAt;

  DocumentModel({
    required this.id,
    required this.name,
    required this.type,
    required this.category,
    required this.ownerId,
    required this.fileUrl,
    required this.fileName,
    required this.mimeType,
    required this.size,
    this.expiresAt,
    required this.status,
    required this.createdAt,
  });

  factory DocumentModel.fromJson(Map<String, dynamic> json) {
    return DocumentModel(
      id: json['_id'] ?? json['id'] ?? '',
      name: json['name'] ?? '',
      type: json['type'] ?? '',
      category: json['category'] ?? '',
      ownerId: json['ownerId'] ?? '',
      fileUrl: json['fileUrl'] ?? '',
      fileName: json['fileName'] ?? '',
      mimeType: json['mimeType'] ?? '',
      size: json['size'] ?? 0,
      expiresAt: json['expiresAt'] != null ? DateTime.parse(json['expiresAt']) : null,
      status: json['status'] ?? 'valid',
      createdAt: json['createdAt'] != null ? DateTime.parse(json['createdAt']) : DateTime.now(),
    );
  }
}
